lib/mosql/streamer.rb

```diff
@@ -73,13 +73,30 @@ def with_retries(tries=10)
           raise if e.kind_of?(Mongo::OperationFailure) && [11000, 11001].include?(e.error_code)
           # Cursor timeout
           raise if e.kind_of?(Mongo::OperationFailure) && e.message =~ /^Query response returned CURSOR_NOT_FOUND/
-          delay = 0.5 * (1.5 ** try)
+          delay = 0.5 * (2 ** try)
           log.warn("Mongo exception: #{e}, sleeping #{delay}s...")
           sleep(delay)
         end
       end
     end
 
+    def tail_with_retries(tries=3)
+      tries.times do |try|
+        begin
+          yield
+        rescue Sequel::PoolTimeout, Sequel::DatabaseConnectionError, Sequel::DatabaseDisconnectError => e
+          # Be less aggressive with the rerties. Every ten minutes three times.
+          delay = (10 * 60) * (try + 1)
+          log.warn("Postgres exception: #{e}, sleeping #{delay}s...")
+          sleep(delay)
+        #Break out of the loop if no exception raised on retrying
+        else
+          break
+        end
+        raise Exception.new("Reached the maximum number of retries") if try == tries - 1
+      end   
+    end
+
```

Comment:
Does this retry behavior fix the problem of hanging on an individual database request? My understanding was that Postgres does not return an exception in some cases. Our issue was that we were waiting for Postgres to return a response and Postgres did not return anything.

If we inserted some records, the connection was dropped, and no exception was thrown, we would still be hanging for ever. I don't think that a timeout on inserts should be specified in the `tail_with_retries` function necessarily, but we should specify a timeout on the request to Postgres.